File: problems 80-89/problem_86.py
# Problem 86
# Cuboid route

import math

EXPECTED = 1000000

def norma_cuad(c1, c2):
	return c1*c1 + c2*c2

def es_cuad_perfecto(d):
	return math.sqrt(d) == round(math.sqrt(d))

# Me da la cantidad de elementos mayores a x en la lista, log n
def cant_mayores(lista, x):
	lo = -1
	hi = len(lista) 
	while lo + 1 < hi:
		m = (lo + hi) // 2
		if lista[m] < x:
			lo = m
		else:
			hi = m

	cant = len(lista) - hi
	return cant


def res(M):
	cuboides = 0
	perfectos = [ [] for _ in range(2*M+1)]

	for base in range(1, 2*M+1):
		for h in range(1, M+1):
			if es_cuad_perfecto( norma_cuad(base, h) ):
				perfectos[base].append(h)

	for a in range(1, M+1):
		for b in range(a, M+1):
			cuboides += cant_mayores(perfectos[a+b], b)

	return cuboides

# binaria en M
lo = 0
hi = 2000
while lo + 1 < hi:
	m = (lo + hi) // 2
	print("Probando M:", m)

	if res(m) < EXPECTED:
		lo = m
	else:
		hi = m

print(hi)

File: problems 80-89/problem_86.cpp
#include <iostream>
#include <vector>
#include <cmath>

using namespace std;

typedef long long ll;

const ll EXPECTED = 1000000;

ll norma_cuad(ll c1, ll c2) {
	return c1*c1 + c2*c2;
}

bool es_cuad_perfecto(ll d) {
	ll a = (ll) sqrt(d);
	return (a * a == d || (a+1) * (a+1) == d);
}

ll cant_mayores(vector<ll> &lista, ll x) {
	ll lo = -1;
	ll hi = lista.size();
	while (lo + 1 < hi) {
		ll m = (lo + hi) / 2;
		if (lista[m] < x)
			lo = m;
		else
			hi = m;
	}
	return lista.size() - hi;
}

ll res(ll M) {
	ll cuboides = 0;
	vector< vector<ll> > perfectos(2*M + 1);

	for (ll base = 1; base <= 2*M; base++) {
		for (ll h = 1; h <= M; h++) {
			if (es_cuad_perfecto( norma_cuad(base, h) ))
				perfectos[base].push_back(h);
		}
	}

	for (ll a = 1; a <= M; a++) {
		for (ll b = a; b <= M; b++) {
			cuboides += cant_mayores(perfectos[a+b], b);
		}
	}

	return cuboides;
}

int main() {
	ll lo = 0;
	ll hi = 2000;
	while (lo + 1 < hi) {
		ll m = (lo + hi) / 2;
		cout << "Probando: " << m << "\n";
		if (res(m) < EXPECTED) 
			lo = m;
		else
			hi = m;
	}

	cout << hi << "\n";
}
